File: Assets/EPICGODDAMNWORLDGENDEMO/PathGen/ChunkCore/ChunkManager.cs
using System.Collections.Generic;
using UnityEngine;

public class ChunkManager
{
    private Chunk[] chunks;

    int gridSize;

    public ChunkManager(int gridSize)
    {
        this.gridSize = gridSize;
        chunks = new Chunk[gridSize * gridSize * gridSize];
    }

    public Chunk GetChunkByLocation(Vector3Int location)
        => chunks[ChunkHelperFunctions.ParseLocationToId(location)];

    public void SetUpEmptyChunks()
    {
        int chunkCount = 0;
        for (int x = 0; x < gridSize; x++)
            for (int y = 0; y < gridSize; y++)
                for (int z = 0; z < gridSize; z++)
                    chunks[chunkCount++] = new(new(x, y, z));
    }

    public Chunk SetChunkTypeByLocation(Vector3Int location, ChunkTypeEnum type)
        => GetChunkByLocation(location).SetChunkType(type);

    public Chunk SetDownHole(Chunk chunk)
    {
        SetChunkTypeByLocation(chunk.location, ChunkTypeEnum.HoleDown);
        Chunk upHole = SetChunkTypeByLocation(new(chunk.location.x, chunk.location.y, chunk.location.z - 1), ChunkTypeEnum.HoleUp);
        return upHole;
    }

    public Chunk SetEndingChunk(Chunk chunk)
    {
        return SetChunkTypeByLocation(chunk.location, ChunkTypeEnum.End);
    }

    public void SetChunkDirections(Chunk chunk, ChunkDesignEnum pathDirections)
        => chunk.SetChunkDesign(pathDirections);

    public void VisualizeChunks()
        => ChunkDebugManager.VisualizeChunks(chunks, gridSize);

    public Chunk[] GetChunks
        => chunks;

    public void EvaluateNeighbors(Vector3Int location, List<DirectionEnum> incoming, List<DirectionEnum> blocked)
    {
        foreach (var (dir, directionEnum) in ChunkHelperFunctions.directions)
        {
            Vector3Int neighborLocation = location + dir;

            if (!ChunkHelperFunctions.IsInsideGrid(neighborLocation))
            {
                blocked.Add(directionEnum);
                continue;
            }

            var neighbor = GetChunkByLocation(neighborLocation);

            if (neighbor.GetChunkType == ChunkTypeEnum.Nothing)
                continue;

            var opposite = ChunkHelperFunctions.GetOpposite(directionEnum);

            if (neighbor.GetOpeningDirections.Contains(opposite))
                incoming.Add(directionEnum);
            else
                blocked.Add(directionEnum);
        }
    }
}

File: JavaMinecraftProject/untitled/src/main/java/com/daan/spelunky3D/pathGen/chunkcore/ChunkManager.java
package com.daan.spelunky3D.pathgen.chunkcore;

import com.daan.spelunky3D.pathgen.debug.ChunkDebugManager;
import com.daan.spelunky3D.pathgen.enums.ChunkDesignEnum;
import com.daan.spelunky3D.pathgen.enums.ChunkTypeEnum;
import com.daan.spelunky3D.pathgen.enums.DirectionEnum;
import com.daan.spelunky3D.pathgen.models.Chunk;
import com.daan.spelunky3D.pathgen.models.Vector3Int;
import com.daan.spelunky3D.pathgen.utils.ChunkHelperFunctions;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.List;

public class ChunkManager {

    private Chunk[] chunks;
    private int gridSize;

    public ChunkManager(int gridSize) {
        this.gridSize = gridSize;
        this.chunks = new Chunk[gridSize * gridSize * gridSize];
    }

    public Chunk getChunkByLocation(Vector3Int location) {
        return chunks[ChunkHelperFunctions.parseLocationToId(location)];
    }

    public void setUpEmptyChunks() {
        int chunkCount = 0;

        for (int x = 0; x < gridSize; x++)
            for (int y = 0; y < gridSize; y++)
                for (int z = 0; z < gridSize; z++)
                    chunks[chunkCount++] = new Chunk(new Vector3Int(x, y, z));
    }

    public Chunk setChunkTypeByLocation(Vector3Int location, ChunkTypeEnum type) {
        return getChunkByLocation(location).setChunkType(type);
    }

    public Chunk setDownHole(Chunk chunk) {
        setChunkTypeByLocation(chunk.location, ChunkTypeEnum.HOLE_DOWN);

        Chunk upHole = setChunkTypeByLocation(
                new Vector3Int(chunk.location.x, chunk.location.y, chunk.location.z - 1),
                ChunkTypeEnum.HOLE_UP
        );

        return upHole;
    }

    public Chunk setEndingChunk(Chunk chunk) {
        return setChunkTypeByLocation(chunk.location, ChunkTypeEnum.END);
    }

    public void setChunkDirections(Chunk chunk, ChunkDesignEnum pathDirections) {
        chunk.setChunkDesign(pathDirections);
    }

    public void visualizeChunks(JavaPlugin plugin) {
        ChunkDebugManager.visualizeChunks(plugin, chunks, gridSize);
    }

    public Chunk[] getChunks() {
        return chunks;
    }

    public void evaluateNeighbors(Vector3Int location,
                                  List<DirectionEnum> incoming,
                                  List<DirectionEnum> blocked) {

        for (DirectionEnum direction : DirectionEnum.values()) {

            if (direction == DirectionEnum.NONE) continue;

            Vector3Int dirVector = ChunkHelperFunctions.directionToVector(direction);
            Vector3Int neighborLocation = location.add(dirVector.x, dirVector.y, dirVector.z);

            if (!ChunkHelperFunctions.isInsideGrid(neighborLocation)) {
                blocked.add(direction);
                continue;
            }

            Chunk neighbor = getChunkByLocation(neighborLocation);

            if (neighbor.getChunkType() == ChunkTypeEnum.NOTHING)
                continue;

            DirectionEnum opposite = ChunkHelperFunctions.getOpposite(direction);

            if (neighbor.getOpeningDirections().contains(opposite)) {
                incoming.add(direction);
            } else {
                blocked.add(direction);
            }
        }
    }
}